### DISCUSSION TITLE
Non-Page Locators

### DISCUSSION THREAD

**Discussion by dstark** [jgm/pandoc#10168] [General]
The Pandoc [manual](https://pandoc.org/MANUAL.html#extension-citations) seems to suggest that non-page locators can be triggered by including an appropriate term, such as `sec.` [This style](https://github.com/citation-style-language/styles/blob/master/society-of-biblical-literature-fullnote-bibliography.csl) includes a section term name (`§` for one, `§§` for multiple). But citations with `sec.` simply print "sec." when Pandoc uses this style, rather than "§" or "§§", although the style is otherwise getting applied properly.

Am I missing something here on the Pandoc side? On the CSL side? Both?

Thanks so much for your thoughts!

**Comment by jgm** [comment]
Hm, this may be a bug. Surprising. I believe this used to work, but now it doesn't seem to work (tried various styles, languages, and locator labels).  I need to look into  this more, maybe there's something I'm forgetting.

**Reply by jgm** [reply]
OK, on further testing, this does seem to be working after all.  The thing to realize is that the locator label is only used for certain types of entries: see line 814 of the csl file.  For example I was able to get it to work by using `report` instead of `article_journal`.

**Reply by dstark** [reply]
Thanks so much, @jgm. With the CSL file I mentioned, if I do a citation with an `annote` variable and a `sec.` identifier, I do get "§" or "§§" as expected in the output since, in that case, the CSL explicitly says what to do when the `annote` variable is present and a section identifier is specified.

What this seems to require, then, is that each possible non-page locator should be explicitly included as a locator option in the CSL in each possible place where it might appear in order for Pandoc to produce the corresponding non-page term output that's specified. Does that sound correct?

If so, might that create what is---within the CSL itself on that specification's own terms---needless duplication? And if so, would a lookup in the `locale` section be worth considering as an enhancement for Pandoc? E.g., if (1) a section (or some other non-page locator) is specified by the user and (2) the CSL `locale` section includes a `term` corresponding to that locator type, render the `term` for that locator type irrespective of whether the CSL for a specific item type includes explicit instructions about how to handle that kind of locator.

My apologies if this is a wrong way of thinking about Pandoc on my part. I'm coming to Markdown and Pandoc from Zotero's built-in word processor integration. And there, Zotero renders locator terms as I'm describing so that, for instance, a `web page` item type with a section locator specified for the citation will automatically render as "§" or "§§" even though the CSL for the `web page` item type doesn't explicitly describe a section locator scenario.

Thanks again for your thoughts!

**Reply by jgm** [reply]
CSL has to specify how to render the locator (a style might even choose not to render it at all!), but it need not do this separately for each type of locator.   Look at this macro from the default CSL pandoc uses, for an example:
```xml
  <macro name="point-locators">
    <choose>
      <if variable="locator">
        <choose>
          <if locator="page" match="none">
            <choose>
              <if type="bill book graphic legal_case legislation motion_picture report song" match="any">
                <choose>
                  <if variable="volume">
                    <group>
                      <text term="volume" form="short" suffix=" "/>
                      <number variable="volume" form="numeric"/>
                      <label variable="locator" form="short" prefix=", " suffix=" "/>
                    </group>
                  </if>
                  <else>
                    <label variable="locator" form="short" suffix=" "/>
                  </else>
                </choose>
              </if>
              <else>
                <label variable="locator" form="short" suffix=" "/>
              </else>
            </choose>
          </if>
          <else-if type="bill book graphic legal_case legislation motion_picture report song" match="any">
            <number variable="volume" form="numeric" suffix=":"/>
          </else-if>
        </choose>
        <text variable="locator"/>
      </if>
    </choose>
  </macro>
```
As you can see, this gives special treatment to the "page" locator, and special treatment to "volume" for some types of documents, but otherwise we get catch-alls like
```xml
        <label variable="locator" form="short" suffix=" "/>
```
(prints label in short form, for any type of locator)
```xml
        <text variable="locator"/>
```
(prints the locator itself).

**Reply by dstark** [reply]
Thanks so much. I'll have a look at this in the CSL the next time I come across it.

**Reply by dstark** [reply]
@jgm, apologies that it's taken me so long to come back around to looking at this more carefully. But yes, after running some tests and trying to retool bits of the CSL, I'm inclined to agree that the difficulties I was finding on this front were a result of the CSL not defining the citation properly—as well as some user error on my part in similar cases where Pandoc outputs the citations perfectly as expected once locators are demarcated with squiggly brackets. Thanks again!

**Comment by badumont** [comment]
So the parsing of the locator depends on the CSL style file, not on the CSL locale file corresponding to the language of the document as stated in the manual?

**Reply by jgm** [reply]
It depends on both.  The style can override some of the terms in the default locale.

**Comment by badumont** [comment]
I see, thank you. So we might need to modify some locators in our source file when we change the CSL style.

**Reply by jgm** [reply]
I guess so.  That might be something we should think about changing in pandoc.  For this purpose it might make sense to use the locale's terms without integrating the style's customizations.
